### DISCUSSION TITLE
Add `node-red admin init` option to help user create settings file

### DISCUSSION THREAD

**Discussion by knolleary** [node-red/designs#53] [Design Proposals]
# Problem

1. to get a settings file, users must first run Node-RED to get the default file copied into the right location. If a user wants to edit their settings file *before* running Node-RED for the first time, they need to get ahold of a settings file from somewhere. We don't make that easy.

2. the default settings file has no security applied. A user has to go through multiple manual steps to setup `adminAuth` - generate password, edit settings file (which may not exist yet... see point 1 above), add the right boilerplate code to the settings file, and so on.

3. there are other settings the user may want to turn on - for example, projects mode.

# Proposed Solution

We add a new command to `node-red-admin` (which is also exposed as `node-red admin`) that will help the user generate the settings file in the right location without having to run Node-RED itself.

The command will be similar in nature to `npm init` or `eslint --init` - by default it will ask the user a series of questions to generate the file. For example:

1. pick location of node-red user dir (with the default being `~/.node-red` as usual)
2. ask for a `credentialSecret` value
3. ask to setup user security - with a suitable warning about leaving unsecured.
4. ask whether to enable projects feature
5. ... any other options that make sense

Need to consider:
1. what if a settings file already exists? We cannot edit it as it is unstructured JavaScript....
2. should there be a 'quiet' mode where cli args can be used to provide answers and/or defaults applied, to avoid any interactive prompts

**Comment by sammachin** [comment]
This sounds like a great idea, editing the settings file is currently quite a barrier to entry for many new users, it would be nice to eventually get to a stage where this could be done from the editor as well, obviously that requires a restart of Node-RED but that would also be useful from the editor when upgrading nodes.

**Comment by knolleary** [comment]
PR raised - https://github.com/node-red/node-red-admin/pull/15

**Comment by dceejay** [comment]
Can it detect Node-RED version and not ask about Monaco if still v1 ?
Is it worth adding any git settings here also - like remote repo etc ?

**Reply by knolleary** [reply]
This is V2 of the tool that is installed with NR V2. So Monaco will be there.

Yes there is an edge case of a user running 1.x decides to install it separately (despite it being built in to Node-RED). But not sure that's worth worrying about.

There will be a separate command for setting up a project. Once that exists, this command can tell the user to either run that, or open the editor to setup their first project.

**Reply by dceejay** [reply]
ok - so it probably doesn't need the (new for 2.0) comment

**Reply by knolleary** [reply]
why not? it's a new feature for 2.0 that users won't necessarily know about. Asking them questions about features they don't know exist could be confusing. Saying its new for 2.0 provides some reassurance they haven't missed something.

**Reply by knolleary** [reply]
... lets discuss the details on the PR against the actual changes.

**Comment by TotallyInformation** [comment]
Would it be possible to add a settings.js compare feature? Some way to compare a live settings.js with the latest default? That is something that is increasingly a pain as the settings file grows in complexity.
